Fix retention edits in WebUI and CLI

dennisjbell opened the pull request:
Two related commits, comprising:

* Changed PUT to PATCH for updating retention policies
  - Fixes https://trello.com/c/Er2n1hnP
* Apply boundry checks to retention plan form
  - Fixes https://trello.com/c/iLk5wUSO and https://trello.com/c/7iCoEiWG

dennisjbell commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Big retention expiries don't return a limit or a proper error message](https://trello.com/c/7iCoEiWG/51-big-retention-expiries-dont-return-a-limit-or-a-proper-error-message)

dennisjbell commented:
![](https://github.trello.services/images/mini-trello-icon.png) [adding a huge named retention policy doesn't return a proper error](https://trello.com/c/iLk5wUSO/54-adding-a-huge-named-retention-policy-doesnt-return-a-proper-error)

dennisjbell commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Cannot rename retention policy on webUI](https://trello.com/c/Er2n1hnP/77-cannot-rename-retention-policy-on-webui)

jhunt commented on web2/htdocs/shield.js:
```diff
@@ -1328,6 +1328,26 @@ function divert(page) { // {{{
+    $form.error('expires', 'missing');
+  } else if (!parseInt(data.days) || parseInt(data.days) < 1 || parseInt(data.days) != data.days) {
+    $form.error('expires', 'invalid');
+  } else if (data.days > 1895489799660) {
```
Let's max this out at 10y.

jhunt commented on web2/htdocs/shield.css:
```diff
@@ -1167,10 +1168,11 @@ form .ctl span.errors {
   float: right;
-  text-align: right;
+  text-align: left;
+  padding-bottom: 6px;
```
Lets leave the styles out of this PR.

jhunt commented on web2/src/js/shield.js:
```diff
@@ -31,6 +31,26 @@ function divert(page) { // {{{
+    $form.error('expires', 'missing');
+  } else if (!parseInt(data.days) || parseInt(data.days) < 1 || parseInt(data.days) != data.days) {
+    $form.error('expires', 'invalid');
+  } else if (data.days > 1895489799660) {
```
Again, 10y.

jhunt commented on web2/htdocs/index.html:
```diff
@@ -1190,6 +1191,7 @@ <h3>Configure a New Retention Policy</h3>[[# {{{ ]]
               <span class="errors">
                 <span data-error="missing">You must specify an expiry of at least 1 day.</span>
                 <span data-error="invalid">This must be a positive, non-zero, whole number.</span>
+                <span data-error="too-big">I'm sorry, the Sun has exploded and taken your data (and the rest of the Earth) with it.<br>Do you realy think you need to retain your data this long?!</span>
```
Let's walk back the snark on this one.

dennisjbell commented:
This now covers the CLI policy and policy templates, as well as the WebUI policy templates in addition to the initial WebUI policies.  Furthermore, the core will also reject anything between the normalized maximums (100 characters for name and 3653 days for expiry) and the maximum possible values for the data types (64k for strings and 2^64 for unsigned integers respectively)

dennisjbell commented:
Additional enhancements/bug fixes were:
* Changed policy-template updates to use PATCH instead of PUT to match policy updates
* Removed Summary from policy-template creation and update help (they weren't supported in the code)
* Updated the API docs

jhunt commented on core/v2.go:
```diff
@@ -1753,11 +1753,21 @@ func (core *Core) v2API() *route.Router {
 			r.Fail(route.Bad(nil, "Retention policy expiry must be greater than 1 day"))
 			return
 		}
+		if in.Expires > 315619200 {
```
Can we make this a calculation, i.e. `366 * 10 * 86400`?

jhunt commented on core/v2.go:
```diff
@@ -1838,6 +1852,10 @@ func (core *Core) v2API() *route.Router {
 				r.Fail(route.Bad(nil, "Retention policy expiry must be greater than 1 day"))
 				return
 			}
+			if in.Expires > 315619200 {
```
Same here, re: calculation

jhunt commented on core/v2.go:
```diff
@@ -3070,12 +3088,20 @@ func (core *Core) v2API() *route.Router {
 			r.Fail(route.Bad(nil, "Retention policy expiry must be greater than 1 day"))
 			return
 		}
+		if in.Expires > 315619200 {
```
also here

jhunt commented on core/v2.go:
```diff
@@ -3151,6 +3181,10 @@ func (core *Core) v2API() *route.Router {
 				r.Fail(route.Bad(nil, "Retention policy expiry must be greater than 1 day"))
 				return
 			}
+			if in.Expires > 315619200 {
```
... and here
